packages/storage-plus/src/snapshot_item.rs

```diff
@@ -0,0 +1,354 @@
+#![cfg(feature = "iterator")]
+
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+use cosmwasm_std::{Order, StdError, StdResult, Storage};
+
+use crate::keys::U64Key;
+use crate::map::Map;
+use crate::snapshot::ChangeSet;
+use crate::{Bound, Item, Strategy};
+
+/// Item that maintains a snapshot of one or more checkpoints.
+/// We can query historical data as well as current state.
+/// What data is snapshotted depends on the Strategy.
+pub struct SnapshotItem<'a, T> {
+    primary: Item<'a, T>,
+
+    // Maps height to number of checkpoints (only used for selected)
+    checkpoints: Map<'a, U64Key, u32>,
+
+    // This stores all changes, by height. Must differentiate between no data written,
+    // and explicit None (just inserted)
+    changelog: Map<'a, U64Key, ChangeSet<T>>,
+
+    // How aggressive we are about checkpointing all data
+    strategy: Strategy,
+}
+
+impl<'a, T> SnapshotItem<'a, T> {
+    /// Usage: SnapshotItem::new(snapshot_names!("foobar"), Strategy::EveryBlock)
+    pub const fn new(
+        storage_key: &'a str,
+        checkpoints: &'a str,
+        changelog: &'a str,
+        strategy: Strategy,
+    ) -> Self {
+        SnapshotItem {
+            primary: Item::new(storage_key),
+            checkpoints: Map::new(checkpoints),
+            changelog: Map::new(changelog),
+            strategy,
+        }
+    }
+
+    pub fn add_checkpoint(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
```

Comment:
We could remove the checkpoint/strategy logic here and just make it "always".
Or refactor this out to share the same logic in both item and map.

It is too complex for me to feel comfortable with copy-and-paste implementation.

Reply:
I'll put this helper in the `snapshot` module, and use it from both structs.